File: 02-network/07-see-it-on-the-wire/java/SynClient.java
// Layer 2 · Topic 7 - Java's contribution to the SYN table.
//
// One HttpClient instance, reused. Its pool is not exposed through the public
// API, so the only way to find out whether it pooled is to count from the
// other end -- which is exactly what this topic is about.
//
//   LAB_URL=http://127.0.0.1:8000/work java -cp /tmp/javabuild SynClient
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

public class SynClient {
    public static void main(String[] args) throws Exception {
        String url = System.getenv().getOrDefault("LAB_URL", "http://127.0.0.1:8000/work");
        int n = Integer.parseInt(System.getenv().getOrDefault("LAB_REQUESTS", "30"));

        HttpClient client = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(5))
                .version(HttpClient.Version.HTTP_1_1)   // match the other five
                .build();
        HttpRequest req = HttpRequest.newBuilder(URI.create(url)).build();

        long t0 = System.nanoTime();
        for (int i = 0; i < n; i++) {
            client.send(req, HttpResponse.BodyHandlers.ofString());
        }
        System.out.printf("one HttpClient instance, %d requests in %d ms%n",
                n, (System.nanoTime() - t0) / 1_000_000L);
    }
}

File: 02-network/07-see-it-on-the-wire/python/syn_client.py
"""
Layer 2 · Topic 7 - Python's contribution to the SYN table.

One reused httpx.Client, LAB_REQUESTS requests. Run by
pools_as_advertised.py, which counts the connections from the server side;
runnable on its own too, against any URL:

    LAB_URL=http://127.0.0.1:8000/work python3 syn_client.py
"""
import os
import sys
import time

URL = os.environ.get("LAB_URL", "http://127.0.0.1:8000/work")
N = int(os.environ.get("LAB_REQUESTS", "30"))

try:
    import httpx
except ImportError:
    sys.exit("httpx not installed: pip install httpx")

t0 = time.perf_counter()
with httpx.Client(timeout=10.0) as client:      # ONE client, kept alive
    for _ in range(N):
        client.get(URL).read()
print(f"httpx.Client reused across {N} requests in "
      f"{(time.perf_counter() - t0) * 1000:.0f} ms")
